plumbing/format/packfile/patch_delta.go

```diff
@@ -73,6 +75,123 @@ func PatchDelta(src, delta []byte) ([]byte, error) {
 	return b.Bytes(), nil
 }
 
+func ReaderFromDelta(h *ObjectHeader, base plumbing.EncodedObject, deltaRC io.ReadCloser) (io.ReadCloser, error) {
+	deltaBuf := bufio.NewReaderSize(deltaRC, 1024)
+	srcSz, err := decodeLEB128ByteReader(deltaBuf)
+	if err != nil {
+		if err == io.EOF {
+			return nil, ErrInvalidDelta
+		}
+		return nil, err
+	}
+	if srcSz != uint(base.Size()) {
+		return nil, ErrInvalidDelta
+	}
+
+	targetSz, err := decodeLEB128ByteReader(deltaBuf)
+	if err != nil {
+		if err == io.EOF {
+			return nil, ErrInvalidDelta
+		}
+		return nil, err
+	}
+	remainingTargetSz := targetSz
+
+	dstRd, dstWr := io.Pipe()
+
+	go func() {
```

Comment:
Why a goroutine instead of returning a reader that calls this code? Could it happen that if you don't read the whole object the goroutine is left behind?

Reply:
If you don't read the whole object - then you will `Close()` the returned `io.ReadCloser`. That reader is `dstRd` a `*io.PipeReader` so closing it will `Close()` the `dstWr *io.PipeWriter`.

This goroutine should only be blocked in a `dstWr.Write(...)` function call - either as part of an `io.Copy` or other `Write`. On `dstRd.Close()` the `dstWr.Write(...)` will fail and return an error, which will then cause `_ = dstWr.CloseWithError(err)` to be called (which does nothing in this case) and then finally `return` killing the go-routine.

If the goroutine gets blocked somewhere else, say L140-L172 - then when it does become unblocked then at the time of writing to dstWr the closed pipe will be seen and the goroutine closed at that point. However, most likely at this point `dstRd` reader is actually going to be blocked in `Read()`. 

(Note: I noticed I had missed a few error-checks and returns below which I have put in.)

Whilst it should be possible to unroll this to get away from the goroutine - that would be much more complicated and require much more state which the pipe makes simpler.